tests/test_component.py

```diff
@@ -518,3 +518,29 @@ async def start(self) -> None:
         await start_component(DummyComponent)
 
     assert "The root component started a service task (servicetask)" in caplog.messages
+
+
+async def test_component_generic_resource(caplog: LogCaptureFixture) -> None:
+    def resource0() -> str:
+        return "foo"
+
+    def resource1() -> int:
+        return 3
+
+    class RootComponent(Component):
+        async def start(self) -> None:
+            add_resource(resource0, types=Callable[[], str])  # type: ignore[arg-type]
+            add_resource(resource1, types=Callable[[], int])  # type: ignore[arg-type]
+            _resource0 = get_resource_nowait(Callable[[], str])  # type: ignore[call-overload]
+            _resource1 = get_resource_nowait(Callable[[], int])  # type: ignore[call-overload]
```

Comment:
Not sure how to get rid of the `type: ignore`.

Reply:
The proper type annotation for the first argument of `get_resource()` or `get_resource_nowait()` is really `typing.TypeForm`, but since [PEP 747](https://peps.python.org/pep-0747/#implicit-typeform-evaluation) is still draft, I don't see any better workaround.